keycloak-claims-mapper/src/main/java/ee/eki/tolkevarav/sso/keycloakserviceprovider/tokenclaimsmapper/TaraSubjectAttributeToPicClaimMapper.java

```diff
@@ -0,0 +1,99 @@
+package ee.eki.tolkevarav.sso.keycloakserviceprovider.tokenclaimsmapper;
+
+import com.google.auto.service.AutoService;
+import org.jboss.logging.Logger;
+import org.keycloak.models.ClientSessionContext;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ProtocolMapperModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.protocol.ProtocolMapper;
+import org.keycloak.protocol.oidc.mappers.*;
+import org.keycloak.provider.ProviderConfigProperty;
+import org.keycloak.provider.ProviderConfigurationBuilder;
+import org.keycloak.representations.IDToken;
+
+import java.util.List;
+
+import static ee.eki.tolkevarav.sso.keycloakserviceprovider.tokenclaimsmapper.TaraParser.TARA_SUBJECT_CLAIM_ATTRIBUTE_KEY;
+import static ee.eki.tolkevarav.sso.keycloakserviceprovider.tokenclaimsmapper.TaraParser.parsePicFromTaraSubClaim;
+import static org.keycloak.provider.ProviderConfigProperty.BOOLEAN_TYPE;
+
+@AutoService(ProtocolMapper.class)
+public class TaraSubjectAttributeToPicClaimMapper extends AbstractOIDCProtocolMapper
```

Comment:
Why was there a need for such custom mapper? IdentityProviderMapper could have done the same thing without any custom code?

Reply:
@thenouan If you mean using "User Attribute" mapper to save user attribute into claim:
* It can not be used to transform the attribute, such as parse the identification code, as this mapper does.
* It can not be used to cause failure conditionally, such as aborting request processing if attribute is not in expected format, as this mapper does (controlled via config param).

Since you mentioned identity provider mappers – afaik mappers on IDPs can not be used to put claims from IDP token into Keycloak token, they can only be used to put claims into the Keycloak user’s group, role or attribute.